LXC backend

earl-warren opened the pull request:
## Description

The  [LXC](https://linuxcontainers.org/) backend is used to **run tests that do not fit the constraints imposed by the docker backend** such as deploying docker, a woodpecker agent and server as well as Forgejo on a systemd capable environment. See [this Woodpecker pipeline for an example of how the LXC backend is used](https://codeberg.org/earl-warren/scripted-configuration/src/branch/wip-lxc/.woodpecker.yml) and the [logs of a successful run](https://ci.dachary.org/earl-warren/scripted-configuration/pipeline/4/3).

It creates a container from scratch on every run. If the tests accidentally damage essential services such as the ssh server, it will not have any impact on the host running the LXC backend. If the same accident happens with the local backend the host itself will be damaged. The **LXC backend provides a robust isolation for each job in the pipeline, which the local backend does not.**

## Implementation details

It is roughly the equivalent of doing the following:

* Creating a LXC container (with lxc-create)
* Running the commands in the container (with lxc-attach)
* Destroying the LXC container (with lxc-destroy)

This backend is inherently insecure (and documented as such). Hardening LXC containers is possible but it makes them no more useful than docker containers.

![image](https://user-images.githubusercontent.com/109468362/214624352-f526c6a5-9484-4049-81eb-9855a1182567.png)

Container images including this PR are available for testing at:

* https://codeberg.org/forgejo-contrib/-/packages/container/woodpecker-forgejo-server/versions
* https://codeberg.org/forgejo-contrib/-/packages/container/woodpecker-forgejo-agent/versions
* https://codeberg.org/forgejo-contrib/-/packages/container/woodpecker-forgejo-cli/versions

## FAQ

* **Why not run LXC inside a Docker container?** Because [it does not work](https://github.com/woodpecker-ci/woodpecker/pull/1565#issuecomment-1413922298). Although it could probably be hacked, that would not be more than that.
* **Why not run the LXC logic from the local backend?** Because it would require significant work and LXC knowledge from each Woodpecker user. Instead they only have to care for their own tests, not to [setup and teardown well isolated LXC containers](https://github.com/woodpecker-ci/woodpecker/pull/1565#issuecomment-1413937609).
* **Why not run the LXC containers as unprivileged to improve security?** Because it does not work with [most templates](https://linuxcontainers.org/lxc/getting-started/#creating-unprivileged-containers-as-a-user). Given the restrictions it imposes, the user is better of using Docker. 
* **Why is apparmor disabled?** Because it cannot be configured [to allow essential operations such as mounting file systems](https://discuss.linuxcontainers.org/t/what-does-security-nesting-true/7156/2)

Ta180m commented on docs/docs/20-usage/20-pipeline-syntax.md:
````diff
@@ -103,6 +103,17 @@ When using the `local` backend, the `image` entry is used to specify the shell,
 +    image: mysql
 ```
 
+When using the `lxc` backend, the `image` entry is used to specify the  `--template`  and `--release`argument to lxc-create, such as  or `debian:bullseye`, that is used to run the commands.
````
Typo, should be "the  `--template` and `--release` argument"

6543 replied:
can you just point out that the backends local, ssh and lxc does use image different and that if you need more info you should read more about that in there dedicated documentation (aka ... link to backends/X) 

and remove this specific case here?

earl-warren commented on pipeline/backend/lxc/lxc.go:
```diff
@@ -0,0 +1,232 @@
+if test {{.Release}} = bullseye ; then
+  sleep 5
+fi
+lxc-attach --name {{.Name}} -- systemctl start dbus
```
`/usr/share/lxc/templates/lxc-download -l` shows a number of images and maybe some of them are not systemd capable. Which one do you have in mind?

earl-warren replied:
I get what you're advocating for, thanks for explaining. Supporting plugins would be useful as a way to mix LXC steps with Docker steps. As it is implemented, the plugins need to be used in a separate pipeline that runs on docker.

Allowing a single pipeline to run steps that require different backends would need architectural changes that are above my paygrade I'm afraid. Unless you already know how to do that? Or maybe I'm looking at this under the wrong angle?

Eragonfr replied:
All alpine images does not support systemctl. It's in the design of Alpine to have a small footprint. This will not work on these images.

earl-warren replied:
This is an outdated thread, the current implementation does not rely on systemd at boot time.

earl-warren commented on pipeline/backend/lxc/lxc.go:
```diff
@@ -0,0 +1,232 @@
+
+lxc-start {{.Name}}
+lxc-wait --name {{.Name}} --state RUNNING
+if test {{.Release}} = bullseye ; then
```
Added

```
#
# This workaround is for a race condition that was fixed in dbus >= v250 which is not
# available in bullseye.
#
```

earl-warren commented on pipeline/backend/lxc/lxc.go:
```diff
@@ -0,0 +1,232 @@
+tee -a /var/lib/lxc/{{.Name}}/config <<'EOF'                                                                                                                   
+security.nesting = true
+lxc.cap.drop =
+lxc.apparmor.profile = unconfined
```
Yes, it is. 

This backend, very much like the local backend, requires a high level of trust. The primary motivation is not secure confinement of the jobs. It is to provide a way to create a fresh systemd capable environment virtually identical to a real machine that can be wiped out and leave no trace behind when it completes. It guards against breakage caused by failed tests but not against malicious jobs.

Maybe it is possible to make this LXC backend very secure but I'm not aware that containers that provide root access can do that effectively.

Eragonfr replied:
You could run rootless containers, running LXC as a user and not as root is a first step. And you could probably create a new user in the LXC container, so it's not root.
But LXC could be described as a glorified chroot. So even though it gives some flexibility, it doesn't provide much security.

Is there a specific reason for disabling apparmor ?

earl-warren replied:
Well put. There is no specific reason to disable apparmor, only it makes it easier because I do not fully understand how to properly harden the underlying LXC container running on the host in a way that would allow nested LXC containers to get all the capabilities they want (including everything down to disabling any protection, but only for nested containers).

Eragonfr replied:
Disabling Apparmor could also be restricted by a variable `AGENT_LXC_DISABLE_APPARMOR`. That way it's optional.

But after some thinking, is it possible to run this with rootless containers ? Do you mount any filsystems in the container ?

earl-warren replied:
Because it cannot be configured [to allow essential operations such as mounting file systems](https://discuss.linuxcontainers.org/t/what-does-security-nesting-true/7156/2). It would be possible to keep it activated and run unprivileged containers. But that would offer no benefit compared to docker containers.

earl-warren commented on pipeline/backend/lxc/lxc.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+var destroyTemplate = template.Must(template.New("destroy").Parse(`#!/bin/sh -x
+umount /var/lib/lxc/{{.Name}}/rootfs/woodpecker
```
Yes, and it goes back to this backend not being suitable for access by random strangers.

earl-warren replied:
I think it is robust although not safe. In addition it comes with integration tests that will help keep it solid. The destroy code was modified to make kill first, umount second so there cannot be processes blocking the umount.

earl-warren commented:
> I think this backend should come with some warning. The whole architecture based on executing shell scripts do not inspire me lots of confidence, far from it.

And it should **not** be trusted, same as the local backend. A documentation page was added to prominently explain this: https://github.com/woodpecker-ci/woodpecker/pull/1565/commits/fd4c5d389aafd7292ed89844c07fa8a789799405#diff-c8d5ca4a41b419284ab219708fd679cb54222ea2771dadc361ca7bf9bf74922e.

earl-warren commented:
@mscherer if there is a consensus that it is the way to go, I'll modify the PR to do that. I would do it right away because it makes sense to me and fits my needs. But I'm not familiar with how decisions / reviews are made yet in Woodpecker and I'd rather wait for a confirmation in case some authoritative disagrees and I have to go in another direction. Is that ok with you?

gapodo commented:
> "and we do not want to use a throwaway VM for that" ( I guess because the Venn diagram of easy options and working options is a empty set ).

I'll try to shed some light on this, LXC likely came up, since it's what Forgejo is using on the infrastructure provided by Codeberg, which was the trigger for the backend, as we are trying to get our integration with woodpecker fully end to end tested, as well as test our own CI deployment (a woodpecker instance) which is running as an LXC guest. Using LXC makes these steps easier to accomplish and more reliable than using docker (timing issues with starting up the tests, enrolling authentication details,...)

Another advantage of LXC (from my personal perspective) compared to a VM is the reduced overhead in reserved memory, as well as not requiring a system that has virtualization capabilities (as is the case with vServers at many VPS hosters).

gapodo commented:
> But anything that can run lxc should also be able to run docker/podman (unless of course specifically configured to not do so). And if the main issue is that test in docker are unreliable due to timeout, I am not sure how lxc would be different in practice. They both expose the same knobs for tuning, so maybe that's what should be investigated first before going on a different stack (who is the same stack underneat).

I understand what you are saying, and I am agreeing with it fully, though I may have not described our use cases for the LXC backend in a way, that is suitable for someone not involved in our work.

Re. the timeouts, it's not timeouts but timing (i.e. start service one, do something start service 2 do something, then do something to service one with the result of service 2,... and once that is done run x)

We have multiple scenarios, where we are planning on using the LXC backend (or are already using it). I'll describe some (those that I am aware off and understand enough to explain) below.

#### Testing our (specific) CI-Setup

At Forgejo our entire CI is hosted within LXC, with woodpecker running from binary (extracted from the docker images) and docker being dedicated to our actual CI runs (running on a RAM-disk for performance during integration testing and being discarded regularly).

This entire setup is scripted to allow us to set up the CI with just a couple of commands (and manually adding the secrets, as we don't want to have them in our setup scripts). Since our setup is scripted, and we want to keep it well tested, there are also tests written, which run the setup and test it. These tests are what we require for accepting changes to the CI setup (we don't manually touch the setup except for adding the secrets).

#### Testing integration with docker / other container runtimes

This one isn't used yet, but on the list of things to implement on our side, though it relies on LXC backend (or the ssh / local backend, LXC being the best one, as cleanup and isolation for parallel runs is achievable more easily).

Using (one of the) "Linux native" backends, allows the setup and use of docker or another container runtime without relying on `dind`, `kind`, `pind`, `cind`,... which helps with testing those, without the potential of side effects from container runtime nesting (LXC has its own issues, though we have those down and are comfortable with them)

#### Testing the woodpecker and Forgejo integration

This is where I'm not fully up to what others have been doing / implementing though it partially relates to the same timing problem (doing things that are not possible via API to services in a specific order, where one needs to issue things to the first service after the second one is running and has "things" setup)

#### unrelated to the work at Forgejo (but related to my day job)

I'm using LXC to run some fairly specific tests against hardware passed through from the host, which profit from being isolated in some parts, but at the same time, share things with the host. And that is where LXC shines compared to docker (though I believe with some workarounds it would be possible to achieve it within docker and podman).

While this isn't run via woodpecker yet, I'd love to get rid of the current way we run those tests and just integrate them into the CI.

-----

A note on at least my mental model on OCI containers (docker, podman, crio, what ever runs your K8s,...) vs LXC, this may come down to me being a Linux sysadmin, that has a highly restricted Kubernetes (OKD/OpenShift, with additional restrictions) platform as his second area of work / expertise (though K8s is starting to overtake the "normal" Linux admin part, as more and more load shifts onto K8s)

* OCI container
  * one process (yeah, sometimes more than exactly one, but the goal is one)
  * highly isolated
  * mount config, potential data, but leave the system alone (if it's not in the image, don't add it unless you are building a new image on top of that image)
  * chroot with comfort features and clearly defined purpose / isolation
* LXC container
  * a "lower overhead" VM / bare metal machine
  * multiple processes / services (though it can also be just one)
  * possibly highly isolated, but can also be used in less restricted ways (think nesting, iptables,...) without requiring full root privileges (`--privileged` in docker)
  * better suited for tasks requiring more setup / changes than an OCI image
  * chroot with comfort features, but more versatile / configurable than an OCI container and more like a VM / bare metal install


Hoping this somewhat explains the intentions / reasoning for using / wanting LXC as a backend.

earl-warren commented:
> @earl-warren I think it would be more productive to discuss the proposal. 

I now realize that this PR assumes the need for a LXC backend is understood. But it needs to be explained. In a attempt to do that, in addition to @gapodo extensive and thorough explanation of the rationale, I added references to an actual real life example to illustrate a context in which it is put to use. It was added to the PR description:

> It is used to run tests that do not fit the constraints impose by the docker backend such as deploying docker, a woodpecker agent and server as well as Forgejo on a systemd capable environment. See [this Woodpecker pipeline for an example of how the LXC backend is used](https://codeberg.org/earl-warren/scripted-configuration/src/branch/wip-lxc/.woodpecker.yml) and the [logs of a successful run](https://ci.dachary.org/earl-warren/scripted-configuration/pipeline/4/3).

earl-warren commented:
That's an original idea and I'm not aware of anyone doing that. Do you have references by any chance to draw inspiration from?

A few weeks ago I looked for LXC enabled docker images before starting this backend but did not find anything maintained or widely used. It seems like an exotic way of doing things. Although it can probably be done. 

I did give it a try because why not:

```
$ docker run --privileged -ti --rm debian:bullseye
# apt-get install -y -qq git libvirt0 libpam-cgfs bridge-utils uidmap dnsmasq-base dnsmasq dnsmasq-utils qemu-user-static
# apt-get install -y -qq lxc
# lxc-create --name="test" --template=debian -- --release bullseye
# tee -a /var/lib/lxc/test/config <<'EOF'
> security.nesting = true
> lxc.cap.drop =
> lxc.apparmor.profile = unconfined
> EOF
# lxc-start --foreground test
lxc-start: test: cgroups/cgfsng.c: __cgfsng_delegate_controllers: 3085 Device or resource busy - Could not enable "+cpuset +cpu +io +memory +hugetlb +pids +rdma +misc" controllers in the unified cgroup "/sys/fs/cgroup//cgroup.subtree_control"
                                                                                               lxc-start: test: start.c: __lxc_start: 1972 Failed to delegate controllers to monitor cgroup
                                        lxc-start: test: tools/lxc_start.c: main: 308 The container failed to start
lxc-start: test: tools/lxc_start.c: main: 313 Additional information can be obtained by setting the --logfile and --logpriority options
```

It is probably just the beginning. 

Docker containers are not designed for that purpose and it feels like it would be a clever hack but not a clean implementation.

earl-warren commented:
As for the local backend, it would mean the burden of maintaining the LXC isolation logic would be on the Woodpecker user. The LXC backend has three shell scripts indeed but it has a little more than that and it makes a difference: the user does not have to worry about isolation or event understanding LXC at all, just about writing their own tests.

earl-warren commented:
> However, the local backend would work and would be perfect for that. Yes, the users have to deal with isolation, but that's already required, per documentation in the PR and per review. If the backend is only running trusted workload, that's already caring about isolation.

Your conclusion is that this PR should be closed as unnecessary? Or are you suggesting something else?

lafriks commented:
I don't think this PR but just need to be fixed to include recommendations here and properly documented about security implications

earl-warren commented:
> I don't think this PR but just need to be fixed to include recommendations here and properly documented about security implications

I'm afraid I don't understand the recommendations that would need implementing. What I do understand from @mscherer is that there is no need for an LXC backend because it should be the responsibility of the Woodpecker user to achieve the same using the local backend.

Or am I missing something?

maxkratz commented:
This backend would be very helpful, for example, to test Ansible "infrastructure as code" that uses Docker Compose for deployment (like [this repo](https://forgejo.dev/forgejo.dev/infrastructure-as-code) I'm working on).
I don't see another way but using LXC apart from dedicating a whole runner/agent VM with Docker installed for this use case.

Eragonfr commented on pipeline/backend/lxc/lxc.go:
```diff
@@ -0,0 +1,286 @@
+  getent hosts wikipedia.org > /dev/null && break
+  sleep 1
+done
+getent hosts wikipedia.org
```
Is this a smart choice to use wikipedia, I understand that they have a really small chance to be down but what will append if the system administrator intentionally disable access to the outside world ?
It is possible to want a localnet only container, for some reason.

earl-warren replied:
I don't know how to do the same reliably using a different method. I'm open to ideas.

Eragonfr replied:
Maybe you can set some sort of `AGENT_PING_HOST` variable on the agent, allowing the wikipedia.org host to be overwritten by the agent admin.

earl-warren replied:
That would work :+1:

Eragonfr replied:
Disable it in LXC, and set the ping address to localhost. That should work.

earl-warren replied:
WOODPECKER_BACKEND_LXC_NETWORK_READY_HOST was added

earl-warren commented:
@mscherer I updated the PR description based on our recent discussions to clearly articulate how and why the LXC backend is useful compared to the local and docker backend.

You advocate that since it is possible to use the local backend to do the same, there is no need for Woodpecker CI to implement a LXC backend. However, by that logic there is also no need to implement the docker backend because the user can run a docker server using the local backend. The LXC backend, just as the Docker backend, relieves the user from the burden of setting up and tearing down a test environment with strong guarantees regarding isolation.

You suggested that the LXC backend could be replaced by a plugin running in privileged Docker containers. While this is possible in theory, I think we are now in agreement that it would be difficult to implement in practice.

Finally you expressed concerns regarding the fact that the LXC backend is no more secure than the local backend and suggested it is documented prominently. The documentation was updated accordingly. In addition explanations regarding the reasons why it cannot be hardened were added to the description of the PR.

Please kindly let me know if this addresses your concerns.

6543 commented:
First look looks good, please update the pull.

I'll have the next days an full review and then it should be just ready to merge

6543 commented:
About the concerns ... did not read them just saw tjere are some ... will alos look into them ...

Eragonfr commented:
Any updates on this ?

6543 commented on docs/docs/30-administration/22-backends/40-lxc.md:
```diff
@@ -0,0 +1,180 @@
+The LXC backend will execute the pipelines on containers that could be used to get access to the host by a malicious actor.
+:::
+
+It is recommended to use this backend only for private setup where the code and
```
can we also add the hint that it's "experimental" ?

earl-warren commented:
This is stalled and I no longer have an incentive to keep working on it.

maxkratz commented:
> This is stalled and I no longer have an incentive to keep working on it.

Is there an alternative within the current stable release of Forgejo to use LXC containers in the CI pipeline?
